src/rdp.c

```diff
@@ -802,6 +823,169 @@ void rdp_draw_filled_rectangle( int tx, int ty, int bx, int by )
     __rdp_ringbuffer_send();
 }
 
+
+/**
+ * @brief Draw a shaded triangle (no alpha support)
+ *
+ * Draws a shaded triangle to the screen. Vertex order is not important.
+ * Before calling this function, make sure that the RDP is set to shaded mode by
+ * calling #rdp_enable_shading.
+ * Based on triangle guide by snacchus on discord: 
+ * https://docs.google.com/document/d/17ddEo61V0suXbSkKP5mY97QxgUnB-QfAjuBIsPiLWko/edit
+ *
+ * Color values are from 0-255.
+ *
+ * @param[in] x1
+ *            Pixel X1 location of triangle
+ * @param[in] y1
+ *            Pixel Y1 location of triangle
+ * @param[in] x2
+ *            Pixel X2 location of triangle
+ * @param[in] y2
+ *            Pixel Y2 location of triangle
+ * @param[in] x3
+ *            Pixel X3 location of triangle
+ * @param[in] y3
+ *            Pixel Y3 location of triangle
+ * @param[in] v1R
+ *            Red value for vertex 1
+ * @param[in] v1G
+ *            Green value for vertex 1
+ * @param[in] v1B
+ *            Blue value for vertex 1
+ * @param[in] v2R
+ *            Red value for vertex 2
+ * @param[in] v2G
+ *            Green value for vertex 2
+ * @param[in] v2B
+ *            Blue value for vertex 2
+ * @param[in] v3R
+ *            Red value for vertex 3
+ * @param[in] v3G
+ *            Green value for vertex 3
+ * @param[in] v3B
+ *            Blue value for vertex 3
+ */
+void rdp_draw_shaded_triangle(float x1, float y1, float x2, float y2, float x3, float y3, float v1R, float v1G, float v1B, 
+                float v2R, float v2G, float v2B, float v3R, float v3G, float v3B)
+{
+    const float to_fixed_11_2 = 4.0f;
+    const float to_fixed_16_16 = 65536.0f;
+
+    if( y1 > y2 ) { SWAP(y1, y2); SWAP(x1, x2); SWAP(v1R, v2R); SWAP(v1G, v2G); SWAP(v1B, v2B); }
+    if( y2 > y3 ) { SWAP(y2, y3); SWAP(x2, x3); SWAP(v2R, v3R); SWAP(v2G, v3G); SWAP(v2B, v3B); }
+    if( y1 > y2 ) { SWAP(y1, y2); SWAP(x1, x2); SWAP(v1R, v2R); SWAP(v1G, v2G); SWAP(v1B, v2B); }
+    
+    int y1f = y1*to_fixed_11_2;
+    int y2f = y2*to_fixed_11_2;
+    int y3f = y3*to_fixed_11_2;
+    
+    y1 = y1f/4.0f;
+    y2 = y2f/4.0f;
+    y3 = y3f/4.0f;
```

Comment:
Based on searching the discord the cause of gaps between triangles (which I was getting, and this appeared to solve) is calculating the slopes using full Y precision, but RDP only using 2bits for Y fraction. I used an implicit cast rather than a rounding function from a combination of laziness and not knowing what kind of rounding would be most correct or desired.

Reply:
This is very interesting. I encountered something like this before as well but wasn't sure whether it was an error on my side or the RDP documentation is inaccurate (recently I tend more and more towards just not trusting SGI's docs on anything). In theory, it should be fine to calculate the slopes in full precision, because the way `xh`, `xm` and `xl` are calculated should account for the fact that the Y-coords are truncated to 2 fractional bits. @anacierdem what did you end up doing in your code?

(Also don't worry about the implicit cast. Truncating to integer should be the right thing here)

Reply:
Doing full precision arithmetic is just fine, though mine was based on 16.16 fixed numbers but I don't think using floats would be an issue either. So we should ideally do not have anything other than xh, xm, xl adjustments. If you are still having seams, I suggest checking color combiner/othe mode settings.

Reply:
If I don't use my toy GL1.0 I wrote, I can make triangles touch in the obvious same-pixel-coordinates way without seams, so I've removed the alteration of the input values in the latest commit.

Reply:
Looks like there's something off about the colors now on hardware, so now something else to dig into.

Reply:
After getting my new capture dongle working, it looks fine on capture, but on my TCL Series-5 it has a green stripe down the left side and white dots at the corners. Capture attached.
![n64shot](https://user-images.githubusercontent.com/187348/139751450-5b3a1491-9ac3-4d28-97c5-59673c25fa18.png)

Reply:
I will try to run this ans see for myself